waitress/channel.py

```diff
@@ -299,18 +303,25 @@ def write_soon(self, data):
             # the async mainloop might be popping data off outbuf; we can
             # block here waiting for it because we're in a task thread
             with self.outbuf_lock:
-                # check again after acquiring the lock to ensure we the
-                # outbufs are not closed
-                if not self.connected:  # pragma: no cover
+                self._flush_outbufs_below_high_watermark()
+                if not self.connected:
                     raise ClientDisconnected
                 if data.__class__ is ReadOnlyFileBasedBuffer:
                     # they used wsgi.file_wrapper
                     self.outbufs.append(data)
                     nextbuf = OverflowableBuffer(self.adj.outbuf_overflow)
                     self.outbufs.append(nextbuf)
+                    self.current_outbuf_count = 0
                 else:
+                    # if we overflowed then start a new buffer to ensure avoid
+                    # this buffer growing unbounded
+                    if self.current_outbuf_count > self.adj.outbuf_high_watermark:
+                        nextbuf = OverflowableBuffer(self.adj.outbuf_overflow)
```

Comment:
Each time we do this, we start back over with a new memory based buffer, until we overflow onto disk, is this something we should worry about?

Reply:
You tell me, seems fine imo. We could always use a file once we start a file for a channel but I don't see the point I guess without some benchmarks.